static/js/components/Periodogram.jsx

```diff
@@ -0,0 +1,549 @@
+import Box from "@material-ui/core/Box";
+import Dygraph from "dygraphs";
+
+import TextLoop from "react-text-loop";
+import { useForm, Controller } from "react-hook-form";
+import { CopyToClipboard } from "react-copy-to-clipboard";
+
+import * as photometryActions from "../ducks/photometry";
+
+const useStyles = makeStyles((theme) => ({
+  copyb: {
+    display: "block",
+    border: "1px solid blue",
+  },
+  dygraphChart: {
+    position: "relative",
+    left: "1px",
+    right: "1px",
+    top: "1px",
+    bottom: "1px",
+    minWidth: "100%",
+    maxHeight: "100%",
+  },
+  media: {
+    maxWidth: "100%",
+    width: "95%",
+  },
+  button: {
+    margin: theme.spacing(1),
+  },
+  paper: {
+    width: "100%",
+    padding: theme.spacing(1),
+    textAlign: "left",
+    color: theme.palette.text.primary,
+  },
+  items: {
+    maxWidth: "100%",
+    width: "10rem",
+    fullWidth: "true",
+    display: "flex",
+    wrap: "nowrap",
+  },
+  spinner: {
+    position: "relative",
+    margin: "auto",
+    width: "50%",
+    fontWeight: "bold",
+    fontSize: "1.25rem",
+    textAlign: "center",
+  },
+  labels: {
+    fontSize: "0.75rem",
+  },
+}));
+
+function transpose(mat) {
+  // transpose matrix
+  return mat[0].map((col, i) => mat.map((row) => row[i]));
+}
+
+function plotdata(xx, yy, me) {
+  const dat = transpose([xx, yy])
+    .map((x, i) => [x[0], [x[1], me[i]]])
+    .sort((a, b) => b[0] - a[0]);
+  // eslint-disable-next-line no-new
+  new Dygraph(document.getElementById("dataplot"), dat, {
+    drawPoints: true,
+    strokeWidth: 0,
+    labels: ["time", "mag"],
+    errorBars: true,
+    valueRange: [Math.max(...yy) + 0.1, Math.min(...yy) - 0.1],
+    dateWindow: [Math.min(...xx) - 6, Math.max(...xx) + 6],
+  });
+}
+
+function plotphased(xx, yy, p, title) {
+  // Create graph with native array as data source
+  const pp = [...xx.map((x) => x % p), ...xx.map((x) => (x % p) + p)];
+  // eslint-disable-next-line no-new
+  new Dygraph(
+    document.getElementById("phased"),
+    transpose([pp, [...yy, ...yy]]),
+    {
+      drawPoints: true,
+      strokeWidth: 0,
+      labels: ["phase", "mag"],
+      valueRange: [Math.max(...yy) + 0.1, Math.min(...yy) - 0.1],
+      dateWindow: [0, 2 * p],
+      title,
+    }
+  );
+}
+
+function plotline(g, x) {
+  const lines = [
+    [1.0, "rgb(255,0,0,0.8)", 3],
+    [0.5, "rgb(200,50,50,0.2)", 1],
+    [3.0, "rgb(200,50,50,0.2)", 1],
+    [2.0, "rgb(200,50,50,0.2)", 1],
+    [4.0, "rgb(200,50,50,0.2)", 1],
+  ];
+
+  g.updateOptions({
+    underlayCallback(canvas) {
+      for (let i = 0; i < lines.length; i += 1) {
+        const loc = Math.log10(x * lines[i][0]);
+        canvas.beginPath();
+        canvas.moveTo(g.toDomXCoord(loc), g.toDomYCoord(0));
+        canvas.lineTo(g.toDomXCoord(loc), g.toDomYCoord(10));
+        [canvas.strokeStyle, canvas.lineWidth] = [lines[i][1], lines[i][2]];
+        canvas.stroke();
+      }
+    },
+  });
+}
+
+function dot(x, y) {
```

Comment:
We should test these functions using something like [Ava](https://github.com/avajs/ava).  We need to introduce a JS tester for the existing plotting code anyway.

Below, I notice e.g. that `mul` has bad behavior for unequal length vectors.

Reply:
I'm using mathjs for vector math now rather than the GLS versions.